enterprise/backend/src/metabase_enterprise/serialization/v2/ingest.clj

```diff
@@ -21,3 +32,83 @@
     [this path]
     "Given one of the `:serdes/meta` abstract paths returned by [[ingest-list]], read in and return the entire
     corresponding entity."))
+
+(extend-type Temporal y.reader/YAMLReader
+             (decode [data]
+               (u.date/parse data)))
+
+(defn- read-timestamps [entity]
+  (->> (keys entity)
+       (filter #(or (#{:last_analyzed} %)
+                    (.endsWith (name %) "_at")))
+       (reduce #(update %1 %2 u.date/parse) entity)))
+
+(defn- map-ish?
+  [x]
+  (or (map? x) (isa? x flatland.ordered.map.OrderedMap)))
+
+(defn- parse-keys
+  "Convert suitable string keys to clojure keywords, ignoring keys with whitespace, etc."
+  [obj]
+  (cond
+    (map-ish? obj)    (m/map-kv (fn [k v]
+                                  [(if (re-matches #"^[0-9a-zA-Z_\./\-]+$" k)
+                                     (apply keyword (str/split k #"/"))
+                                     k)
+                                   (parse-keys v)])
+                                obj)
+    (sequential? obj) (mapv parse-keys obj)
+    :else             obj))
+
+(defn- strip-labels
+  [hierarchy]
+  (mapv #(dissoc % :label) hierarchy))
+
+(defn- ingest-file
+  "Reads an entity YAML file and clean it up (eg. parsing timestamps)
+  The returned entity is in \"extracted\" form, ready to be passed to the `load` step."
+  [file]
+  (-> file
+      (yaml/from-file false)
+      parse-keys
+      read-timestamps))
+
+(def ^:private legal-top-level-paths
+  #{"actions" "collections" "databases" "snippets"}) ; But return the hierarchy without labels.
+
+(defn- ingest-all [^File root-dir]
+  ;; This returns a map {unlabeled-hierarchy [original-hierarchy File]}.
+  (into {} (for [^File file (file-seq root-dir)
+                 :when      (and (.isFile file)
+                                 (let [rel (.relativize (.toPath root-dir) (.toPath file))]
+                                   (-> rel (.subpath 0 1) (.toString) legal-top-level-paths)))
+                 ;; TODO: only load YAML once.
+                 :let [hierarchy (serdes/path (ingest-file file))]]
```

Comment:
This is a significant change: all files are scanned and the hierarchy is loaded from the YAML file. This means each YAML file is loaded twice, with the code as it is currently written, but I don't think this is actually a big performance problem. There's no noticeable impact when I run it.